Neo4j.Driver/Neo4j.Driver/Internal/Session.cs

```diff
@@ -68,97 +98,104 @@ protected override void Dispose(bool isDisposing)
-                var resultBuilder = new ResultBuilder(statement, statementParameters, () => _connection.ReceiveOne(), _connection.Server);
-                lock (_txSyncLock)
-                {
-                    EnsureCanRunMoreStatements();
-                    _connection.Run(statement, statementParameters, resultBuilder);
-                    _connection.Send();
-                }
-                return resultBuilder.PreBuild();
-            });
+            EnsureSessionIsOpen();
+            EnsureNoOpenTransaction();
+            DisposeSessionResult();
         }
 
-        public ITransaction BeginTransaction(string bookmark = null)
+
+        /// <summary>
+        ///  This method will be called back by <see cref="ResultBuilder"/> after it consumed result
+        /// </summary>
+        private void RunResultCleanupCallback()
         {
-            return TryExecute(() =>
-            {
-                lock (_txSyncLock)
-                {
-                    EnsureCanRunMoreStatements();
-                    _transaction = new Transaction(_connection, _transactionCleanupAction, _logger, bookmark);
-                }
-                return _transaction;
-            });
+            Throw.ArgumentNullException.IfNull(_sessionRunResult, nameof(_sessionRunResult));
+            Throw.ArgumentNullException.IfNull(_connection, nameof(_connection));
+
+            CleanRunResultResources();
         }
 
-        public string LastBookmark { get; private set; }
+        private void CleanRunResultResources()
+        {
+            LastBookmark = null;
+            _sessionRunResult = null;
 
-        private void EnsureCanRunMoreStatements()
+            // always try to close connection used by the result too
+            _connection?.Dispose();
+            _connection = null;
+        }
+
+        /// <summary>
+        /// Called back in <see cref="Transaction.Dispose"/>
+        /// </summary>
+        private void TransactionCleanupCallback()
         {
-            EnsureConnectionIsHealthy();
-            EnsureNoOpenTransaction();
-            EnsureSessionIsOpen();
+            Throw.ArgumentNullException.IfNull(_transaction, nameof(_transaction));
+            Throw.ArgumentNullException.IfNull(_connection, nameof(_connection));
+
+            LastBookmark = _transaction.Bookmark;
+            _transaction = null;
+
+            // always dispose connection used by the transaction too
+            _connection.Dispose();
+            _connection = null;
         }
 
-        private void EnsureSessionIsOpen()
+        /// <summary>
+        /// Clean any session.run result reference.
+        /// If session.run result is not fully consumed, then pull full result into memory.
+        /// </summary>
+        /// <exception cref="ClientException">If error when pulling result into memory</exception>
+        private void DisposeSessionResult()
         {
-            if (!_isOpen)
+            if (_sessionRunResult != null)
```

Comment:
Yes, but I keep a reference to result just to keep track that whose connection I am disposing now. Would like to attach some debugging tool later for better debugging.